pkg/controller/uniteddeployment/uniteddeployment_update.go

```diff
@@ -0,0 +1,151 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package uniteddeployment
+
+import (
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+
+	appsv1alpha1 "github.com/openkruise/kruise/pkg/apis/apps/v1alpha1"
+	"github.com/openkruise/kruise/pkg/util"
+)
+
+func (r *ReconcileUnitedDeployment) manageSubsets(ud *appsv1alpha1.UnitedDeployment, nameToSubset map[string]*Subset, nextReplicas, nextPartitions map[string]int32, currentRevision, updatedRevision *appsv1.ControllerRevision, subsetType subSetType) (updateErr error) {
+	_, err := r.manageSubsetProvision(ud, nameToSubset, nextReplicas, nextPartitions, currentRevision, updatedRevision, subsetType)
+	if err != nil {
+		return fmt.Errorf("fail to manage Subset provision: %s", err)
+	}
+
+	return
+}
+
+func (r *ReconcileUnitedDeployment) manageSubsetProvision(ud *appsv1alpha1.UnitedDeployment, nameToSubset map[string]*Subset, nextReplicas, nextPartitions map[string]int32, currentRevision, updateRevision *appsv1.ControllerRevision, subsetType subSetType) (sets.String, error) {
+	expectedSubsets := sets.String{}
+	gotSubsets := sets.String{}
+
+	for _, subset := range ud.Spec.Topology.Subsets {
+		expectedSubsets.Insert(subset.Name)
+	}
+
+	for subsetName := range nameToSubset {
+		gotSubsets.Insert(subsetName)
+	}
+	klog.V(4).Infof("UnitedDeployment %s/%s has subsets %v, expects subsets %v", ud.Namespace, ud.Name, gotSubsets.List(), expectedSubsets.List())
+
+	var creates []string
+	for _, expectSubset := range expectedSubsets.List() {
+		if gotSubsets.Has(expectSubset) {
+			continue
+		}
+
+		creates = append(creates, expectSubset)
+	}
+
+	var deletes []string
+	for _, gotSubset := range gotSubsets.List() {
+		if expectedSubsets.Has(gotSubset) {
+			continue
+		}
+
+		deletes = append(deletes, gotSubset)
+	}
+
+	var errs []error
+	// manage creating
+	if len(creates) > 0 {
+		// do not consider deletion
+		klog.V(0).Infof("UnitedDeployment %s/%s needs creating subset (%s) with name: %v", ud.Namespace, ud.Name, subsetType, creates)
+		creatdChan := make(chan string, len(creates))
+		defer close(creatdChan)
+		for _, subset := range creates {
+			creatdChan <- subset
+		}
+
+		revision := currentRevision.Name
+		if updateRevision != nil {
+			revision = updateRevision.Name
+		}
+
+		var createdNum int
+		var createdErr error
+		createdNum, createdErr = util.SlowStartBatch(len(creates), slowStartInitialBatchSize, func() error {
```

Comment:
Can we add index parameter to func() type?  It looks weird to use channel to pass parameter.

Reply:
done